EIPS/eip-9999.md

```diff
@@ -0,0 +1,285 @@
+    validator_index: ValidatorIndex
+    data: PayloadAttestationData
+    signature: BLSSignature
+```
+
+```python
+class IndexedPayloadAttestation(Container):
+    attesting_indices: List[ValidatorIndex, PTC_SIZE]
+    data: PayloadAttestationData
+    signature: BLSSignature
+```
+
+```python
+class SignedExecutionPayloadHeader(Container):
+    message: ExecutionPayloadHeader
+    signature: BLSSignature
+```
+    
+```python
+class ExecutionPayloadEnvelope(Container):
+    payload: ExecutionPayload
+    builder_index: ValidatorIndex
+    beacon_block_root: Root
+    blob_kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
+    payload_withheld: boolean
+    state_root: Root
+```
+
+```python
+class SignedExecutionPayloadEnvelope(Container):
+    message: ExecutionPayloadEnvelope
+    signature: BLSSignature
+```
+
+The `BeaconState` container is modified with the addition of:
+- `latest_block_hash`, of type `Hash32`, to track the blockhash of the last execution payload in the blockchain. 
+- `latest_full_slot`, of type `Slot`, to track the latest slot in which an execution payload was included. 
+- `latest_widthdrawals_root` of type `Root` to track the hash tree root of the latests withdrawals deducted in the beacon chain when processing a `SignedBeaconBlock`. 
+
+State transition logic is modified by: 
+
+- A new beacon state getter `get_ptc` returns the PTC members for a given slot. 
+- `get_attesting_indices` is modified to not return the beacon committee members that are included in the PTC
+- `process_withdrawals` is modified as follows. Withdrawals are obtained directly from the beacon state instead of the execution payload. They are deducted from the beacon chain. The beacon state `latest_widthdrawals_root` is updated with the HTR of this list. The next execution payload MUST include withdrawals matching the `state.latest_withdrawals_root`. 
+- `process_execution_payload` is removed from `process_block`. Instead a new function `process_execution_payload_header` is included, this function validates the `SignedExecutionPayloadHeader` included in the `BeaconBlockBody` and transfers the payment from the builder's balance to the proposer. 
+- `process_deposit_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- `process_withdrawal_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- `process_consolidation_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- A new `process_payload_attestation` is added to `process_operations`, this function validates the payload timeliness attestations broadcast by the PTC members. 
+- `process_execution_payload` is now called as a separate helper when receiving a `SignedExecutionPayloadEnvelope` on the P2P layer. This function in particular checks that the HTR of the resulting beacon state coincides with the committed one in the payload envelope. 
```

Comment:
I think `HTR of the resulting beacon state` can be more clearly expressed. Perhaps something like `HTR of the cached builder header in beacon state`. Right now it sounds like we are comparing the payload envelope with the beacon state.

Reply:
no, I really meant HTR of the resulting beacon state, there's a state transition in this function.

Reply:
Ah thanks. This is my misunderstanding. I thought you meant to check the header matches the payload. Resolved